asclepias_broker/api/ingestion.py

```diff
@@ -0,0 +1,454 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2018 CERN.
+#
+# Asclepias Broker is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Relationshps ingestion functions."""
+
+import uuid
+from typing import Tuple
+
+from invenio_db import db
+from sqlalchemy.orm import aliased
+
+from ..models import Group, GroupM2M, GroupMetadata, GroupRelationship, \
+    GroupRelationshipM2M, GroupRelationshipMetadata, GroupType, Identifier, \
+    Identifier2Group, Relation, Relationship, Relationship2GroupRelationship
+
+
+def merge_group_relationships(group_a, group_b, merged_group):
+    """Merge the relationships of merged groups A and B to avoid collisions.
+
+    Groups 'group_a' and 'group_b' will be merged as 'merged_group'.
+    This function takes care of moving any duplicate group relations, e.g.:
+
+    If we have 4 relations:
+        A Cites X
+        B Cites X
+        Y Cites A
+        Y Cites B
+    and we merge groups A and B, we also need to squash the first two and last
+    two relations together:
+
+        {AB} Cites X
+        Y Cites {AB}
+
+    before we can perform the actual marging of A and B. Otherwise we will
+    violate the unique constraint. We do that by removing the duplicate
+    relationships (only one of each duplicate pair), so that we can later
+    execute and UPDATE.
+    """
+    # Determine if this is an Identity-type group merge
```

Comment:
it would be useful here to say 'as opposed to ...' to say what other kinds of group merge there is.